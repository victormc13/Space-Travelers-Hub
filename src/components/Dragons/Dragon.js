import PropTypes from 'prop-types';

function Dragon({ dragon, onReserve }) {
  const badgeStyle = dragon.reserved
    ? { padding: '2px 10px' }
    : { padding: '0' };

  const buttonStyle = dragon.reserved
    ? { background: '#fff', color: '#000' }
    : { background: '#007bff', border: 'none' };

  return (
    <div className="dragon">
      <img src={dragon.flickr_images[0]} alt={dragon.name} className="dragon-img" />
      <div className="dragon-detail">
        <h2>{dragon.name}</h2>
        <p>
          <span style={badgeStyle} className="badge">
            {dragon.reserved ? 'Reserved' : ''}
          </span>
          {dragon.description}
        </p>
        <button
          style={buttonStyle}
          className="btn-reservation"
          type="button"
          onClick={() => onReserve(dragon.id)}
        >
          {dragon.reserved ? 'Cancel Reservation' : 'Reserve dragon'}
        </button>
      </div>
    </div>
  );
}

Dragon.propTypes = {
  dragon: PropTypes.shape({
    id: PropTypes.string.isRequired,
    flickr_images: PropTypes.arrayOf(PropTypes.string).isRequired,
    name: PropTypes.string.isRequired,
    reserved: PropTypes.bool.isRequired,
    description: PropTypes.string.isRequired,
  }).isRequired,
  onReserve: PropTypes.func.isRequired,
};

export default Dragon;
