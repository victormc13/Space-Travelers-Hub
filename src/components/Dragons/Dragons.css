.rockets-container {
  display: grid;
  gap: 2rem;
}

.dragon {
  display: flex;
  gap: 3rem;
}

.dragon-img {
  width: 30%;
}

.dragon-detail {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  justify-content: flex-start;
  align-items: flex-start;
}

.dragon-detail h2 {
  font-size: 2rem;
  color: #25292d;
}

.badge {
  background-color: #18a2b8;
  color: #fff;
  border-radius: 4px;
  margin-right: 0.6rem;
}

.btn-reservation {
  background-color: #007bff;
  color: #fff;
  padding: 1rem 1rem;
  font-size: 1.2rem;
  border-radius: 5px;
  cursor: pointer;
}
