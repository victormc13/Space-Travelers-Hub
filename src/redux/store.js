import { configureStore } from '@reduxjs/toolkit';
import rocketReducer from './rockets/rocketsSlice';
import missionReducer from './missions/missionSlice';
import dragonReducer from './dragon/dragonsSlice';

const store = configureStore({
  reducer: {
    rockets: rocketReducer,
    missions: missionReducer,
    dragons: dragonReducer,
  },
});

export default store;
