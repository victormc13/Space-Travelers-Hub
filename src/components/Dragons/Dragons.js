import React, { useEffect } from 'react';
import './Dragons.css';
import { useDispatch, useSelector } from 'react-redux';
import { fetchDragons, reserve } from '../../redux/dragon/dragonsSlice';
import Dragon from './Dragon';

function Dragons() {
  const dispatch = useDispatch();
  const { dragons, status } = useSelector((state) => state.dragons);

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchDragons());
    }
  }, [status, dispatch]);

  return (
    <div className="dragons-container">
      {status === 'loading' && <div>Loading...</div>}
      {status === 'failed' && <div>Error: Something went wrong</div>}
      {status === 'succeeded'
        && dragons.map((dragon) => (
          <Dragon
            key={dragon.id}
            dragon={dragon}
            onReserve={(dragonId) => dispatch(reserve(dragonId))}
          />
        ))}
    </div>
  );
}

export default Dragons;
