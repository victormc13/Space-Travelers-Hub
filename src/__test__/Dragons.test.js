import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/store';
import Dragons from '../components/Dragons/Dragons';

test('Dragons component renders correctly', () => {
  const { container } = render(
    <Provider store={store}>
      <Dragons />
    </Provider>,
  );

  expect(container).toMatchSnapshot();
});
