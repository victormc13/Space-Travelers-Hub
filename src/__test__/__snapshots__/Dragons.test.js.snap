// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dragons component renders correctly 1`] = `
<div>
  <div
    class="dragons-container"
  >
    <div>
      Loading...
    </div>
  </div>
</div>
`;
